soc/riscv/riscv-ite/it8xxx2/ilm.c

```diff
@@ -0,0 +1,180 @@
+ * address range non-cacheable (which is appropriate because Flash has high latency but RAM is
+ * essentially the same speed as cache).
+ */
+
+extern const uint8_t __ilm_flash_start[];
+extern const uint8_t __ilm_flash_end[];
+extern uint8_t __ilm_ram_start[];
+extern uint8_t __ilm_ram_end[];
+
+#define ILM_BLOCK_SIZE 0x1000
+BUILD_ASSERT((ILM_BLOCK_SIZE & (ILM_BLOCK_SIZE - 1)) == 0, "ILM_BLOCK_SIZE must be a power of two");
+
+#define FLASH_BASE CONFIG_FLASH_BASE_ADDRESS
+#define RAM_BASE   CONFIG_SRAM_BASE_ADDRESS
+#define RAM_SIZE   CONFIG_SRAM_SIZE
+
+#define ILM_NODE DT_NODELABEL(ilm)
+
+#define SCARH_ENABLE	 BIT(3)
+#define SCARH_ADDR_BIT19 BIT(7)
+
+/*
+ * SCAR registers contain 20-bit addresses in three registers, with one set
+ * of SCAR registers for each ILM block that may be configured.
+ */
+struct scar_reg {
+	/* Bits 0..7 of address; SCARnL */
+	uint8_t l;
+	/* Bits 8..15 of address; SCARnM */
+	uint8_t m;
+	/* Bits 16..18 and 19 of address, plus the enable bit for the entire SCAR; SCARnH */
+	uint8_t h;
+};
+
+struct ilm_config {
+	/* Points to the contiguous set of RVILMCRn registers */
+	volatile uint8_t *ilm_control_regs;
+	volatile struct scar_reg *scar_regs[15];
+};
+
+bool it8xxx2_is_ilm_configured(void)
+{
+	return device_is_ready(DEVICE_DT_GET(ILM_NODE));
+}
+
+static bool __maybe_unused is_block_aligned(const void *const p)
+{
+	return ((uintptr_t)p & (ILM_BLOCK_SIZE - 1)) == 0;
+}
+
+static int it8xxx2_configure_ilm_block(const struct ilm_config *const config, void *ram_addr,
+				       const void *flash_addr, const size_t copy_sz)
+{
+	if ((uintptr_t)ram_addr < RAM_BASE)
+		return -EFAULT; /* Not in RAM */
+	const int dirmap_index = ((uintptr_t)ram_addr - RAM_BASE) / ILM_BLOCK_SIZE;
+
+	if (dirmap_index >= ARRAY_SIZE(config->scar_regs))
+		return -EFAULT; /* Past the end of RAM */
+	BUILD_ASSERT((FLASH_BASE & GENMASK(19, 0)) == 0,
+		     "Flash is assumed to be aligned to SCAR register width");
+	if (((uintptr_t)flash_addr - FLASH_BASE) & ~GENMASK(19, 0))
+		return -EFAULT; /* Address doesn't fit in the SCAR */
+
+	LOG_DBG("Enabling ILM%d %p -> %p, copy %d", dirmap_index, flash_addr, ram_addr, copy_sz);
+
+	volatile uint8_t *const rvilmcr = &config->ilm_control_regs[dirmap_index / 8];
+	volatile struct scar_reg *const scar = config->scar_regs[dirmap_index];
+	const int cr_mask = BIT(dirmap_index % 8);
+
+	int irq_key = irq_lock();
+
+	/* Invalidate any possibly-existing mapping */
+	*rvilmcr &= ~cr_mask;
+	/* Copy block contents from flash into RAM */
+	memcpy(ram_addr, flash_addr, copy_sz);
+	/* Enable ILM block */
+	*rvilmcr |= cr_mask;
+	/* Program SCAR */
+	scar->l = (uintptr_t)flash_addr & GENMASK(7, 0);
+	scar->m = ((uintptr_t)flash_addr & GENMASK(15, 8)) >> 8;
+
+	uint8_t scarh_value = ((uintptr_t)flash_addr & GENMASK(18, 16)) >> 16;
+
+	if ((uintptr_t)flash_addr & BIT(19))
+		scarh_value |= SCARH_ADDR_BIT19;
+	scar->h = SCARH_ENABLE | scarh_value;
```

Comment:
You need to clear SCARH_ENABLE bit here, because 0b = enable scratch sram.

Reply:
Changed to match the previous behavior of the flash driver, setting `SCARH_ENABLE` before memcpy and clearing it on final configuration. Does this mean that we can't also place data in the same 4k block of RAM (so we need to pad `__ilm_ram_end`), or is it still safe to do DLM access to this memory while `SCARH_ENABLE` is clear?

I'd like to be able to segment a given 4k region of RAM by use so mutable data can be allocated in the same 4k block to minimize wasted space, but if ILM and DLM access to a given block are mutually exclusive I guess we need to pad.

Reply:
Since you have added pad after “__ilm_flash_end” to prevent text sections redirect to RAM incorrectly.
It is okay to do DLM access in the same 4K block.